src/daft-plan/src/planner.rs

```diff
@@ -145,41 +166,59 @@ pub fn plan(logical_plan: &LogicalPlan) -> DaftResult<PhysicalPlan> {
-                    ));
-                    Ok(PhysicalPlan::Flatten(Flatten::new(split_op.into())))
-                }
+            let input_partition_spec = input_physical.partition_spec();
+            let input_num_partitions = input_partition_spec.num_partitions;
+            let num_partitions = num_partitions.unwrap_or(input_num_partitions);
+            let repartitioned_plan = match scheme {
+                PartitionScheme::Unknown => match num_partitions.cmp(&input_num_partitions) {
+                    Ordering::Greater => {
+                        // Split input partitions into num_partitions.
+                        let split_op = PhysicalPlan::Split(Split::new(
+                            input_physical.clone(),
+                            input_num_partitions,
+                            num_partitions,
+                        ));
+                        PhysicalPlan::Flatten(Flatten::new(split_op.into()))
+                    }
+                    Ordering::Less => {
+                        // Coalesce input partitions into num_partitions.
+                        PhysicalPlan::Coalesce(Coalesce::new(
+                            input_physical.clone(),
+                            input_num_partitions,
+                            num_partitions,
+                        ))
+                    }
+                    Ordering::Equal => {
+                        // # of output partitions == # of input partitions, so we drop the redundant repartition.
+                        return Ok(input_physical.as_ref().clone());
+                    }
+                },
                 PartitionScheme::Random => {
                     let split_op = PhysicalPlan::FanoutRandom(FanoutRandom::new(
-                        input_physical,
-                        *num_partitions,
+                        input_physical.clone(),
+                        num_partitions,
                     ));
-                    Ok(PhysicalPlan::ReduceMerge(ReduceMerge::new(split_op.into())))
+                    PhysicalPlan::ReduceMerge(ReduceMerge::new(split_op.into()))
                 }
                 PartitionScheme::Hash => {
                     let split_op = PhysicalPlan::FanoutByHash(FanoutByHash::new(
-                        input_physical,
-                        *num_partitions,
+                        input_physical.clone(),
+                        num_partitions,
                         partition_by.clone(),
                     ));
-                    Ok(PhysicalPlan::ReduceMerge(ReduceMerge::new(split_op.into())))
+                    PhysicalPlan::ReduceMerge(ReduceMerge::new(split_op.into()))
                 }
                 PartitionScheme::Range => unreachable!("Repartitioning by range is not supported"),
+            };
+            let repartitioned_partition_spec = repartitioned_plan.partition_spec();
+            if (input_num_partitions == 1 && repartitioned_partition_spec.num_partitions == 1)
```

Comment:
QQ: Why do we perform dropping of repartition in the construction of `repartitioned_plan` for the [PartitionScheme::UNKNOWN case](https://github.com/Eventual-Inc/Daft/pull/1540/files#diff-abc819095f64baaa47a1d26c1633c44897e88ec2fb4bff32357b6aef2226aedcR191-R194), but for all other cases the "dropping" happens in this if-block?

Could be nice to consolidate logic in `repartitioned_plan and have it be per-PartitionScheme type. In that case we would trivially drop the `PartitionScheme::Range` case I guess since we are leaving that `unimplemented!()`